Allow `"*"` in `kubernetes_users`

boxofrad opened the pull request:
Fixes #58274.

changelog: Support setting `"*"` in role `kubernetes_users`

boxofrad committed:
Allow `"*"` in `kubernetes_users`

tigrato commented on lib/kube/proxy/forwarder.go:
```diff
@@ -2066,24 +2089,28 @@ func computeImpersonatedPrincipals(kubeUsers, kubeGroups map[string]struct{}, he
 	}
 
-	if len(impersonateGroups) == 0 {
+	if len(impersonateGroups) == 0 && !hasUserWildcard {
```
This check should not exist here. Users and groups should be independent

tigrato replied:
you can let the function reach the end. no need to return early

boxofrad replied:
I think I might've misunderstood your comment. What did you have in mind for where the check should be?

We presumably can't check at role creation time — because the user might have one role with the wildcard and another with `kubernetes_groups`, right?

creack commented:
If we do it for users, wouldn't it make sense to allow `*` for groups as well?

boxofrad committed:
Add tests for `kubernetes_users` wildcard

boxofrad committed:
Document the `kubernetes_users` wildcard support

boxofrad committed:
Remove the coupling of users and groups

boxofrad commented:
Hey folks 👋🏻 

Following a Slack conversation with @tigrato, I've removed the behavior where we would not automatically impersonate any groups if `kubernetes_users` contained a wildcard. This constraint made sense in the context of Argo CD, but is too restrictive for the general case. I've added a warning to the Kubernetes Access docs instead, and will update #58229 accordingly.

He also suggested that we combine `computeImpersonatedPrincipals` with `fillDefaultKubePrincipalDetails`. I think this makes sense, but I'd rather keep the diff as small as possible, especially as I'm new to this part of the codebase.

Similarly, @creack's suggestion of supporting `*` in `kubernetes_groups` also makes sense - let's maybe revisit it in another PR?

boxofrad committed:
Fix broken test
